### PR TITLE
[JENKINS-24805] Mask all bindings with asterix

### PR DESCRIPTION
**Pull request by idstein**
- Order bindings by largest to shortest string to cover substrings
- Fix an issue if secret is partially duplicated, e.g. "android"
  "androiddebugkey" which results in "***_" and "**_*debugkey"

### CODE REVIEW

**Comment by idstein** [comment]
Some tests showed that this is a breaking change to Stapler framework. If we want to have that checkbox, it must be be capable of migrating existing jobs to this this new SecretBuildWrapprr using the newInstance method on its Descriptor.

**Comment by jglick** [comment]
I suppose this is trying to solve [JENKINS-24805](https://issues.jenkins-ci.org/browse/JENKINS-24805)?

File: src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly [hunk 1]
```diff
@@ -29,4 +29,9 @@ THE SOFTWARE.
             <f:repeatableHeteroProperty field="bindings" hasHeader="true"/>
         </f:block>
     </f:section>
+    <f:advanced>
+    	<f:entry title="Show bindings" field="showBindings" description="If checked, all secret variable bindings will be clear-text.">
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly]: Why is this even an option? I would just delete it. `BindingStep` does not offer such an option, and AFAIK no one has ever asked for it.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 2]
```diff
@@ -71,6 +87,22 @@
...
     }
+    
+    @Override
+    public OutputStream decorateLogger(AbstractBuild build, OutputStream logger)
+    		throws IOException, InterruptedException, RunnerAbortedException {
+    	if(showBindings)
+    		return super.decorateLogger(build, logger);
+    	
+    	Map<String,String> overrides = new HashMap<String,String>();
+        List<MultiBinding.Unbinder> unbinders = new ArrayList<MultiBinding.Unbinder>();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: Why are you collecting these? The result is unused.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 3]
```diff
@@ -71,6 +87,22 @@
...
+        for (MultiBinding<?> binding : bindings) {
+            MultiBinding.MultiEnvironment environment = binding.bind(build, build.getWorkspace(), null, null);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: Do not call `bind` as it will actually do things, like copy files, which `setUp` already did. Just call `binding.variables()`, as `makeSensitiveBuildVariables` does.

File: src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly (hunk 1 above)
Comment (idstein) [inline src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly]: I though that it would be really useful for debugging especially, if you want to see if there are any errors related to your password having a typo or similar or a path that is not accessible at all.

The old behaviour would is to show all bindings, so that option would show the same behaviour as before.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java (hunk 2 above)
Comment (idstein) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: As referred in the next review comment, I thought it would be necessary to call the unbind step after calling the bind step, but that is not necessary anymore after changing the bind step to not do any actual binding, when called with an invalid launcher and workspaces (is null at this stage, but may change in future).

**Comment by idstein** [comment]
In the future Jenkins core must not only mask sensitive variables for Ant and Maven but also provide a proper interface for all sort of runners such as Gradle, Shell or Batch.

For now, using the bindings with an invalid workspaces may produce a proper variable value set. This in turn can be used to decorate the logger and avoid the lacking of passwords. However, this may also mask false positive in the logs such as an arbitrary log statement using the secrete variable value.

File: src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly (hunk 1 above)
Comment (daniel-beck) [inline src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly]: I agree with Jesse. This looks a lot like YAGNI. If there's actual demand (and not just one person), this can always be added later.

Comment (idstein) [inline src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly]: If you do not need it, you don't have to check it. It's unit tested and does not do any harm due to it's simple logic.

I do not see how this contradicts the Jenkins contribution guidance.

**Comment by idstein** [comment]
Any update on this open pull request?

File: src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly (hunk 1 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly]: > does not do any harm

Except that once added, it cannot be removed.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 4]
```diff
@@ -125,15 +127,25 @@
...
+				return o1.length() - o2.length();
+			}
+		};
 
         Filter(Collection<String> secrets) {
             StringBuilder b = new StringBuilder();
-            for (String secret : secrets) {
+            List<String> hiddenSecrets = new ArrayList<String>(secrets);
+            Collections.sort(hiddenSecrets, StringLengthComparator);
+            Collections.reverse(hiddenSecrets);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: Why would you not just invert the comparator?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 5]
```diff
@@ -125,15 +127,25 @@
...
+			public int compare(String o1, String o2) {
+				return o1.length() - o2.length();
+			}
+		};
 
         Filter(Collection<String> secrets) {
             StringBuilder b = new StringBuilder();
-            for (String secret : secrets) {
+            List<String> hiddenSecrets = new ArrayList<String>(secrets);
+            Collections.sort(hiddenSecrets, StringLengthComparator);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: Is this intended as some sort of performance optimization for the regexp, or what?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java [hunk 6]
```diff
@@ -49,7 +49,9 @@
     }
 
     @Override public SingleEnvironment bindSingle(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException {
-        FileCredentials credentials = getCredentials(build);
+        if(workspace==null)
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java]: Huh? `workspace` should be considered `@Nonnull`. Anyway what does this have to do with the rest of the PR?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 7]
```diff
@@ -29,28 +29,44 @@
...
 import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
 
 @SuppressWarnings({"rawtypes", "unchecked"}) // inherited from BuildWrapper
 public class SecretBuildWrapper extends BuildWrapper {
 
     private final List<? extends MultiBinding<?>> bindings;
+    private boolean showBindings;
 
+    public SecretBuildWrapper(List<? extends MultiBinding<?>> bindings, boolean showBindings) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: Unnecessary, delete.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 8]
```diff
@@ -29,28 +29,44 @@
...
+        this.bindings = bindings;
+        this.showBindings = showBindings;
+    }
+    
     @DataBoundConstructor public SecretBuildWrapper(List<? extends MultiBinding<?>> bindings) {
         this.bindings = bindings;
+        showBindings = true;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: Why? The setter will be called immediately afterwards anyway.

And if you meant to retain former behavior for projects loaded from disk for some reason (I do not recommend it), this does not do it. You would need to add a `readResolve`.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 9]
```diff
@@ -71,6 +87,28 @@
...
+    
+    @Override
+    public OutputStream decorateLogger(AbstractBuild build, OutputStream logger)
+    		throws IOException, InterruptedException, RunnerAbortedException {
+    	if(showBindings)
+    		return super.decorateLogger(build, logger);
+    	
+    	Map<String,String> overrides = new HashMap<String,String>();
+        for (MultiBinding<?> binding : bindings) {
+            MultiBinding.MultiEnvironment environment = binding.bind(build, null, null, null);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: No, do not do this. `bind` should be called only once, during `setUp`.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java (hunk 5 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: Or something to do with substrings? Should be done as a separate PR. Nothing to do with the rest of this PR.

File: src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly (hunk 1 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper/config.jelly]: > Especially, as sensitive variables are only masked for Ant or Maven based builds. That means any command line like build step such as Gradle may fail with a hidden whitespace exception via its parameters.

What does this mean?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java (hunk 6 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/FileBinding.java]: I guess it is a side effect of calling `bind` with null values, which you should not do.

**Comment by escoem** [comment]
Hello @idstein , I'm interested in finishing the PR so if you aren't going to do and you agree I will address all stuff commented by @jglick and @daniel-beck

**Comment by escoem** [comment]
Hi, I'm ready for the new PR.

**Comment by idstein** [comment]
My bad, I didn't find time to work on this. Of course, you can go ahead and finish this issue.

**Comment by escoem** [comment]
Great @idstein , the new PR will be focused only in the asterisk.

**Comment by jglick** [comment]
@abayer said he would try to do this the right way.
